pecos/core/utils/mmap_util.hpp

```diff
@@ -0,0 +1,530 @@
+            // Get file size
+            struct stat file_stat;
+            fstat(fd, &file_stat);
+            mmap_size_ = file_stat.st_size;
+            if (mmap_size_ <= 0) {
+                throw std::runtime_error("Memory mapped file size should be positive, got: " + std::to_string(mmap_size_));
+            }
+
+            // Creat mmap
+            int mmap_flags = MAP_SHARED;
+            if (pre_load) { // pre-fault all pages to load them into memory
+                mmap_flags |= MAP_POPULATE;
+            }
+            mmap_ptr_ = mmap(NULL, mmap_size_, PROT_READ, mmap_flags, fd, 0);
+            if (mmap_ptr_ == MAP_FAILED) {
+                throw std::runtime_error("Memory map failed.");
+            }
+        }
+
+        /* Free memory-mapped region */
+        void free_mmap_() {
+            if (mmap_ptr_) {
+                auto res = munmap(mmap_ptr_, mmap_size_);
+                if (res == EINVAL) {
+                    throw std::runtime_error("Free memory map failed.");
+                }
+                mmap_ptr_ = nullptr;
+                mmap_size_ = 0;
+            }
+        }
+}; // end class MmapStoreLoad_
+
+
+/*
+ * Module Private Class to save a series of data, along with metadata and signature into a memory-mapped file.
+ */
+class MmapStoreSave_ {
+    public:
+        /* Constructor to open a file and save to given filename */
+        MmapStoreSave_(const std::string& file_name) {
+            fp_ = fopen(file_name.c_str(), "wb");
+            if (!fp_) {
+                throw std::runtime_error("MmapStoreSave: Open file failed.");
+            }
+        }
+
+        /* Destructor to save metadata and close opened file after save */
+        ~MmapStoreSave_() { finalize_(); }
+
+        /* Save a simple-serializable vector(array) */
+        template<class T, class TT=T, if_simple_serializable<TT> = true>
+        void fput_multiple(const T* src, const uint64_t n_elements) {
+            // Data size
+            uint64_t size = sizeof(T) * n_elements;
+            // Pad previous block with dummy bytes to make sure new block aligned
+            auto n_bytes_to_pad = metadata_.get_n_bytes_to_pad(size);
+            const char dummy = '\0';
+            for(auto i = 0u; i < n_bytes_to_pad; i++){
+                file_util::fput_one(dummy, fp_);
+            }
+            // Save array
+            metadata_.append(size);
+            file_util::fput_multiple<T>(src, n_elements, fp_);
+        }
+
+        /* Save one element */
+        template<class T, class TT=T, if_simple_serializable<TT> = true>
+        void fput_one(const T& src) { // Call fput_multiple
+            fput_multiple<T>(&src, 1u);
+        }
+
+    private:
+        MmapMetadata_ metadata_; // Metadata
+        FILE* fp_; // File pointer
+
+        /* Save metadata to end of file after all data blocks have been saveed.
+         * Also, file pointer is closed at the end.
+         */
+        void finalize_() {
```

Comment:
If we don't plan to reveal this method and only use it once in the destructor, I would suggest to move the content into the destructor itself.